kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/channel/Coordinator.java

```diff
@@ -302,21 +329,37 @@ private Snapshot latestSnapshot(Table table, String branch) {
 
   private Map<Integer, Long> lastCommittedOffsetsForTable(Table table, String branch) {
     Snapshot snapshot = latestSnapshot(table, branch);
-    while (snapshot != null) {
-      Map<String, String> summary = snapshot.summary();
-      String value = summary.get(snapshotOffsetsProp);
-      if (value != null) {
-        TypeReference<Map<Integer, Long>> typeRef = new TypeReference<Map<Integer, Long>>() {};
-        try {
-          return MAPPER.readValue(value, typeRef);
-        } catch (IOException e) {
-          throw new UncheckedIOException(e);
-        }
-      }
-      Long parentSnapshotId = snapshot.parentId();
-      snapshot = parentSnapshotId != null ? table.snapshot(parentSnapshotId) : null;
+
+    if (snapshot == null) {
+      return Map.of();
+    }
+
+    Iterable<Snapshot> branchAncestry =
+        SnapshotUtil.ancestorsOf(snapshot.snapshotId(), table::snapshot);
+    return lastCommittedOffsets(branchAncestry);
+  }
+
+  private Map<Integer, Long> lastCommittedOffsets(Iterable<Snapshot> snapshots) {
+    return Streams.stream(snapshots)
+        .filter(Objects::nonNull)
+        .filter(snapshot -> snapshot.summary().containsKey(snapshotOffsetsProp))
+        .map(snapshot -> snapshot.summary().get(snapshotOffsetsProp))
+        .map(this::parseOffsets)
+        .findFirst()
+        .orElseGet(Map::of);
+  }
+
+  private Map<Integer, Long> parseOffsets(String value) {
+    if (value == null) {
+      return Map.of();
+    }
```

Comment:
if there is a key `snapshotOffsetsProp` but values are empty / null is it a illegal state ?

Reply:
It's not a valid state but I don't think it'd be possible to reach this point if the summary somehow has the offsets key with no values. Summary values are interpreted as string maps as per the spec, and I'd think the parser [implementation would fail](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/SnapshotParser.java#L156) when trying to build the in-memory representation.

 Unless I'm missing something, I'd probably simplify this a bit, writing it with the assumption that there is some structured value associated with the key, and just keep 357-361 so we handle the failure case where the offsets have different types or a different structure than expected.

Reply:
It shouldn't be the case that it's null, but someone might think they can reset offsets by setting it to null.  We should protect against that, which is what happens here.